internal/admission/validator.go

```diff
@@ -630,3 +632,55 @@ func (m *managerClientConsumerGetter) ListAllConsumers(ctx context.Context) ([]k
 	}
 	return consumers.Items, nil
 }
+
+func (validator KongHTTPValidator) ValidateCustomEntity(ctx context.Context, entity kongv1alpha1.KongCustomEntity) (bool, string, error) {
+	// If the spec.contollerName does not match the ingress class name,
+	// ignore it as it is not controlled by the controller.
+	// REVIEW: should we also match the "ingress class" annotation here?
+	if !validator.ingressClassMatcher(&metav1.ObjectMeta{
+		Annotations: map[string]string{
+			annotations.IngressClassKey: entity.Spec.ControllerName,
+		},
+	}, annotations.IngressClassKey, annotations.ExactClassMatch) {
+		validator.Logger.V(util.DebugLevel).Info("Skipped validation because the controller name does not match", "controller_name", entity.Spec.ControllerName)
+		return true, "", nil
+	}
+
+	// REVIEW: should we validate group/kind of `spec.parentRef` here because we only want it to be attached to plugins?
+
+	fields, err := kongstate.RawConfigToConfiguration(entity.Spec.Fields.Raw)
+	if err != nil {
+		return false, fmt.Sprintf(ErrTextCustomEntityFieldsUnmarshalFailed, err), nil
+	}
+
+	schemaService, hasClient := validator.AdminAPIServicesProvider.GetSchemasService()
+	// Skip validation on Kong gateway if we do not have available client.
+	if !hasClient {
+		validator.Logger.V(util.DebugLevel).Info("Skipped because no schema service available")
+		return true, "", nil
+	}
+
+	// Retrieve the schema of the entity from Kong gateway.
+	entityType := entity.Spec.EntityType
+	schema, err := schemaService.Get(ctx, entityType)
+	if err != nil {
+		validator.Logger.V(util.DebugLevel).Info("Failed to get schema of entity", "entity_type", entityType, "error", err)
+		return false, fmt.Sprintf(ErrTextCustomEntityGetSchemaFailed, entityType, err), nil
+	}
+
+	// Extract field definitions from the response of getting the schema of the entity.
+	s := kongstate.ExtractEntityFieldDefinitions(schema)
+	// Fill in fields that are "foreign" and required if there are such fields.
+	for fieldName, field := range s.Fields {
+		// Fill in an arbitrary UUID if the entity requires a reference to a foreign key.
+		// This format of foreign reference in entities should satisfy most of foreign fields in entities(services/routes/consumers/...).
```

Comment:
This isn't obvious from context in the admission validator; we should mention why we need a fake ID here:

```suggestion
		// Fill in an arbitrary UUID if the entity requires a reference to a foreign key.
		// This format of foreign reference in entities should satisfy most of foreign fields in entities(services/routes/consumers/...).
		// The admission webhook needs a placeholder value because the actual value isn't present in the KongCustomEntity spec.
		// It will be derived from the resource(s) used by the parent KongPlugin during translation.
```

Reply:
done.